### PR TITLE
Allow TransformerEncoder and TransformerDecoder to accept 0-dim batch sized tensors.

### PR DESCRIPTION
**Pull request by v0dro**
This issue fixes a part of https://github.com/pytorch/pytorch/issues/12013, which is summarized concretely in  https://github.com/pytorch/pytorch/issues/38115.

This PR allows TransformerEncoder and Decoder (alongwith the inner `Layer` classes) to accept inputs with 0-dimensional batch sizes.

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -5031,7 +5031,10 @@ def multi_head_attention_forward(
     #
     q = q.contiguous().view(tgt_len, bsz * num_heads, head_dim).transpose(0, 1)
     if static_k is None:
-        k = k.contiguous().view(-1, bsz * num_heads, head_dim).transpose(0, 1)
+        if bsz == 0:
+            k = k.contiguous().view(num_heads, 0, head_dim).transpose(0, 1)
+        else:
+            k = k.contiguous().view(-1, bsz * num_heads, head_dim).transpose(0, 1)
```
Comment (jbschlosser) [inline torch/nn/functional.py:5037]:
```suggestion
        k = k.contiguous().view(k.shape[0], bsz * num_heads, head_dim).transpose(0, 1)
```

File: torch/nn/functional.py [hunk 2]
```diff
@@ -5040,7 +5043,10 @@ def multi_head_attention_forward(
             f"expecting static_k.size(2) of {head_dim}, but got {static_k.size(2)}"
         k = static_k
     if static_v is None:
-        v = v.contiguous().view(-1, bsz * num_heads, head_dim).transpose(0, 1)
+        if bsz == 0:
+            v = v.contiguous().view(num_heads, 0, head_dim).transpose(0, 1)
+        else:
+            v = v.contiguous().view(-1, bsz * num_heads, head_dim).transpose(0, 1)
```
Comment (jbschlosser) [inline torch/nn/functional.py:5049]:
```suggestion
        v = v.contiguous().view(v.shape[0], bsz * num_heads, head_dim).transpose(0, 1)
```

**Review by jbschlosser — COMMENTED** [review]
Looking pretty good! Some minor comments below- I think you can simplify the change as suggested

File: test/test_nn.py [hunk 3]
```diff
@@ -13207,6 +13207,65 @@ def test_Bilinear_empty(self, device):
...
+            self._test_module_empty_input(transformer_encoder, input, check_size=False)
+
+    @onlyOnCPUAndCUDA
+    def test_TransformerDecoderLayer_empty(self, device):
+        for batch_first, memory_shape, tgt_shape in [(True, (0, 10, 512), (0, 20, 512)),
+                                                     (False, (10, 0, 512), (20, 0, 512))]:
+            memory = torch.rand(*memory_shape, device=device)
+            tgt = torch.rand(*tgt_shape, requires_grad=True, device=device)
+            decoder_layer = nn.TransformerDecoderLayer(d_model=512, nhead=8, batch_first=batch_first).to(device)
+            out = decoder_layer(tgt, memory)
```
[the remark this answers is not in the document]
Comment (v0dro) [reply test/test_nn.py:13236]: Its hard to have a `check_size` for the multiple inputs since you can't assume which tensor in the inputs is to checked with. However I've added a new function that deals with multiple inputs.

File: test/test_nn.py [hunk 4]
```diff
@@ -12380,6 +12380,20 @@ def _test_module_empty_input(self, module, inp, check_size=True):
                 self.assertEqual(p.grad, torch.zeros_like(p.grad))
         self.assertEqual(inp.grad, torch.zeros_like(inp))
 
+    def _test_module_empty_inputs(self, module, inputs):
```
Comment (jbschlosser) [inline test/test_nn.py:12383]: To avoid the duplication, would it work to change the signature to `_test_module_empty_inputs(self, module, *inputs)` and combine this with `_test_module_empty_input`?

Can call it like `self._test_module_empty_inputs(transformer_decoder, src, target)` for multiple inputs

File: test/test_nn.py [hunk 5]
```diff
@@ -12370,7 +12370,7 @@ def _test_GroupNorm_cuda_half(self):
 
     def _test_module_empty_input(self, module, inp, check_size=True):
         inp.requires_grad_(True)
-        out = module(inp)
+        out = module(*inp)
```
Comment (jbschlosser) [inline test/test_nn.py:12373]: I think this should be reverted

**Review by jbschlosser — COMMENTED** [review]
Implementation looks great! One minor testing maintenance comment

File: test/test_nn.py [hunk 6]
```diff
@@ -13225,6 +13239,50 @@ def test_Bilinear_empty(self, device):
...
+            transformer_decoder = nn.TransformerDecoder(decoder_layer, num_layers=6).to(device)
+            self._test_module_empty_inputs(transformer_decoder, [tgt, memory])
+
+    @onlyOnCPUAndCUDA
+    def test_Transformer_empty(self, device):
+        for batch_first, src_shape, tgt_shape in [(True, (10, 0, 512), (20, 0, 512))]:
+            transformer_model = nn.Transformer(nhead=16, num_encoder_layers=12).to(device)
+            src = torch.rand(*src_shape, requires_grad=True, device=device)
+            tgt = torch.rand(*tgt_shape, requires_grad=True, device=device)
+            self._test_module_empty_inputs(transformer_decoder, [src, tgt])
```
Comment (jbschlosser) [inline test/test_nn.py:13284]: `transformer_decoder` -> `transformer_model` to fix failing checks here

**Review by jbschlosser — COMMENTED** [review]
Just a couple testing tweaks to finish this out :)

**Review by jbschlosser — APPROVED** [review]
LGTM

**Comment by jeffdaily** [comment]
This PR may have broken ROCm CI.  PR #63354 attempts the revert to see if ROCm CI passing status is restored.

**Comment by jeffdaily** [comment]
- https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.2-py3.6-trigger/14074/
- https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.2-py3.6-trigger/14082/
- https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.2-py3.6-trigger/14302/
- https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.2-py3.6-trigger/14347/

ROCm CI test1 fails throughout this PR.
